ql/pricingengines/futures/discountingperpetualfuturesengine.cpp

```diff
@@ -0,0 +1,258 @@
+                Real time = timeGrid[i];
+                fundingRateGrid[i] = fundingRateInterp(time);
+                interestRateDiffGrid[i] = interestRateDiffInterp(time);
+            }
+            
+            if (arguments_.fundingType == PerpetualFutures::AHJ_alt) {
+                Real ratio = 1.;
+                Size i;
+                for (i = 0; i < timeGrid.size() - 1; ++i) {
+                    Real time = timeGrid[i];
+                    Real nextTime = timeGrid[i + 1];
+                    ratio = effForCurve->discount(nextTime) / effForCurve->discount(time)
+                          / effDomCurve->discount(nextTime) * effDomCurve->discount(time);
+                    fundingRateGrid[i] *= ratio;
+                    interestRateDiffGrid[i] *= ratio;
+                }
+                // for i = timeGrid.size() - 1
+                fundingRateGrid[i] *= ratio;
+                interestRateDiffGrid[i] *= ratio;
+            }
+            auto productIRDiff = [timeGrid, fundingRateGrid](Size i) {
+                Real ret = 1.;
+                for (Size j = 0; j <= i; ++j) {
+                    ret /= 1. + fundingRateGrid[j];
+                }
+                return ret;
+            };
+            Real sum = 0.;
+            std::vector<Real> df_dom, df_for;
+            for (Size i = 0; i < timeGrid.size() - 1; ++i) {
+                Real time = timeGrid[i];
+                sum += productIRDiff(i) * (fundingRateGrid[i] - interestRateDiffGrid[i])
+                       * effForCurve->discount(time) / effDomCurve->discount(time);
+                df_dom.push_back(effDomCurve->discount(time));
+                df_for.push_back(effForCurve->discount(time));
+            }
+            Size iLast = timeGrid.size() - 1;
+            Real timeLast = timeGrid[iLast];
+            Real productIRDiffLast = productIRDiff(iLast);
+            Real fundingRateGridLast = fundingRateGrid[iLast];
+            Real interestRateDiffGridLast = interestRateDiffGrid[iLast];
+            Real dt = 1.e-4;
+            Real domRateLast = (log(effDomCurve->discount(timeLast)) - log(effDomCurve->discount(timeLast + dt))) / dt;
+            Real forRateLast = (log(effForCurve->discount(timeLast)) - log(effForCurve->discount(timeLast + dt))) / dt;
+
+            // for t > maxT_, assume flat extrapolaiton on all rates
+            Real lastTerm = productIRDiffLast
+                             * (fundingRateGridLast - interestRateDiffGridLast)
+                             * effForCurve->discount(timeLast) / effDomCurve->discount(timeLast);
+            Real timeStep = (timeGrid.back() - timeGrid.front()) / (timeGrid.size() - 1);
+            Real ratio =
+                1. / (1. + fundingRateGridLast) * exp(-timeStep * (forRateLast - domRateLast));
+            sum += lastTerm / (1. - ratio);
+            factor = sum;
+
+        } else {
+        // continuous-time case
+            Real timeIntegral = 0.;
```

Comment:
The Linux and MacOS builds fail because thay flag this as an unused variable.  Should we delete it, or should it be used somewhere?

Reply:
I have pushed a commit to delete it. Thank you for pointing it out.